app/javascript/src/Policies/components/CustomPolicy.jsx

```diff
@@ -0,0 +1,163 @@
+    'version': '[0.0.1]',
+    'configuration': {
+      'type': 'object',
+      'properties': {
+        'property': {
+          'description': 'A description of your property',
+          'type': 'string'
+        }
+      }
+    }
+  },
+  directory: '',
+  id: 0
+}
+
+function CSRFToken ({win = window}: {win?: any}): React.Node {
+  const getMetaContent = meta => win.document.head.querySelector(`meta[name~=${meta}][content]`).content
+  const props = {
+    name: getMetaContent('csrf-param'),
+    value: getMetaContent('csrf-token'),
+    type: 'hidden'
+  }
+  return (
+    <input {...props} />
+  )
+}
+
+function Editor ({onChange, code}: {onChange: (Object) => void, code: Object}): React.Node {
+  const [ state, setState ] = useState({valid: true, code: toJson(code)})
+
+  const checkConfiguration = (schema) => {
+    if (!schema.configuration) throw new Error('Policy configuration not found')
+  }
+
+  const onCodeChange = (editor, metadata, code) => {
+    try {
+      const schema = fromJson(code)
+      checkConfiguration(schema)
+      onChange(schema)
+      setState({ valid: true, code })
+    } catch (err) {
+      console.warn(err)
+      setState({ valid: false, code })
+    }
+  }
+
+  const [ icon, cls, hiddenClass, errorClass ] = state.valid
+    ? [ 'check', 'valid', 'is-hidden', '' ]
+    : [ 'times', 'invalid', '', 'CustomPolicy-error' ]
+
+  return (
+    <div className={`${errorClass} panel panel-default`}>
+      <div className="panel-heading">
+        <i className={`${cls} fa fa-${icon}`} />
+        {' JSON Schema'}
+      </div>
+      <CodeMirror
+        value={state.code}
+        onChange={onCodeChange}
+        autoCursor={false}
+        options={CM_OPTIONS}
+      />
+      <div className={hiddenClass}> There's an error in the JSON Schema</div>
+    </div>
+  )
+}
+
+function CustomPolicyForm ({policy, onChange}: {policy: Policy, onChange: OnChange}): React.Node {
+  const isNewPolicy = !(policy.id && policy.id !== 0)
+  const action = (isNewPolicy) ? '/p/admin/registry/policies/' : `/p/admin/registry/policies/${policy.id}`
+  return (
+    <form action={action} method="post">
+      <label>
+        Directory:
+        <input type="text" name="directory" value={policy.directory} onChange={onChange} disabled={!isNewPolicy} />
+      </label>
+      <input type="hidden" name="id" value={policy.id} disabled={true} />
+      <input name="schema" type="hidden" value={JSON.stringify(policy.schema)} />
+      {(!isNewPolicy) ? <input name="_method" type="hidden" value='put' /> : ''}
+      <input type="submit" />
+      <CSRFToken />
+    </form>
+  )
+}
+
+function Form ({initialPolicy}: {initialPolicy: Policy}): React.Node {
+  const [policy, setPolicy] = useState(initialPolicy)
+  const onSchemaEdited = (policy: Policy) => (schema: Object) => setPolicy({...policy, ...{schema}})
+  const handleChange = (policy: Policy) => (ev: InputEvent) => setPolicy({...policy, ...{[ev.target.name]: ev.target.value}})
```

Comment:
I really think these pseudo closures are redundant, they're wrapping the same `policy`. Each time `policy` changes, the component will rerender and the callbacks will be created again, with the new value.

If the goal is using the previous state, then a function can be passed to `setState`:
```js
const onSchemaEdited = (schema: Object) => setPolicy(prevPolicy => {...prevPolicy, ...{schema}})
```
But I think this is not needed either although I'm not 100% positive.

Reply:
look @josemigallas, no more closures!

Reply:
and for the record, those weren't "pseudo" closure, but real ones.